### ISSUE TITLE
how to turn off timestamps in exported metrics

### ISSUE THREAD

**Issue by muzammil360** [google/cadvisor#2526]
Hi. I am using cadvisor with pushgateway. It so happens that cadvisor exports timestamps along with metrics. This is a problem with pushgateway as it doesn't accept metrics with timestamps. Following is the error returned if I push the metrics using curl

```
pushed metrics are invalid or inconsistent with existing metrics: pushed metrics must not have timestamps
```
It seems that pushgateway doesn't seem to honor time stamps. Is there anyway I can turn them off in cadvisor? Or if you know an option where pushgateway will ignore the time stamp, even that is acceptable.

**NOTE**: I know that pushing cadvisor metrics to pushgateway is an anti-pattern but for the time being, I don't have much option.

**Comment by dashpole** [comment]
Removing timestamps makes many metrics unusable, since we collect metrics out-of-band.  I don't think we should support turning off timestamps.

**Comment by muzammil360** [comment]
Thanks for the reply.

Is there any way i can integrate cadvisor metrics with pushgateway?
Pushgateway rejects the metrics for the sole reason that it has timestamps.

The authors of pushgateway are adamant that metrics should not have
timestamps. Even if timestamp is ignored, it will work for me but they send
back 400 error.

Any ideas how can i make it work?

**Comment by dashpole** [comment]
I'm not aware of any ways to make that work.

**Comment by SuperQ** [comment]
Timestamps in Prometheus metrics cause more problems than just the pushgateway. For example, exposing timetsamps breaks staleness handling. This causes containers that have been removed to be still visible in the data for 5 minutes.

I don't know why you think the metrics are "unusable" without timestamps. For Prometheus monitoring, we expect no timestamps for metrics for almost every use case. The metric scrape is intended to be "When Prometheus last saw this data".

Exposing timestamps is causing problems for Kubernetes users.

CC @paulfantom @brian-brazil @roidelapluie @brancz

**Comment by dashpole** [comment]
At scrape time, the metrics returned may be up to 15 seconds old.  Rates (e.g. cpu usage) didn't really work without timestamps. 

We might be able to drop timestamps if the metrics are collected "on demand", which was added a few releases back: https://github.com/google/cadvisor/issues/1989.

**Comment by SuperQ** [comment]
On demand would be much preferred, but exposing timestamps, even a few minutes stale, is just fine. I'm not sure what you did to determine "didn't work", but what cAdvisor is doing right now is much worse.

Exposing timestamps is a violation of Prometheus metrics best practices and should not be done. The linked Kubernetes operator issue describes this well.

**Comment by dashpole** [comment]
See https://github.com/google/cadvisor/pull/2059.  CPU rates can be inaccurate by +- 50%, which most users consider "unusable".

Can you link to the best practices documentation which states that timestamps should never be done?

**Comment by SuperQ** [comment]
Pre-compute rates shouldn't be exposed on a Promteheus endpoint. They're not useful for Prometheus users. cAdvisor should only expose the raw container CPU counters.

I will try and find the documentation on timestamps.

**Comment by dashpole** [comment]
These aren't precomputed.  These are the raw CPU counters.  That is exactly the problem.  If the "real" time at which a counter is collected differs significantly from the scrape time, the rate won't be correct.  For example, if we are collecting (in the background) and caching:

t0: 0
t10: 10
t25: 25

If we then scrape at t0, t9, t11, t24, t26 (I know prometheus does regular intervals, but this problem still occurs, just not as dramatically), we get rates:

t0-t9: 0/9 = 0
t9-t11: 10/2 = 5
t11-t24: 0/2 = 0
t24-t26: 13/2 = 6.5

The correct rate is 1 for the entire interval, but prometheus would graph numbers that are dramatically incorrect.

From my understanding, it _is_ a best practice to perform collection at scrape time, and thus not expose timestamps.  However, given that we do not perform collection at scrape time, it seems like we must attach timestamps so that rate computations are correct.  Collecting all metrics at once causes problems when running a non-trivial number of containers (e.g. 100), which is why we don't do that by default.  However, we did recently add the ability to trigger collection at scrape time.  For users that are running low pod density, this could be a good option, and we could remove timestamps in that case.

Keep in mind that prometheus server isn't the only consumer of cAdvisor metrics.  Not attaching timestamps for cached metrics would break rate calculations for all backends, so doing that across the board doesn't seem like a viable option.

**Comment by SuperQ** [comment]
I will have to look at the other issue more closely, but what you're describing is not how Prometheus does calculations.

I will have to go over the linked issue, but the conclusions of the linked PR are incorrect. There is not enough information in that PR to show what's going on for real. They have one graph with 6 hours data, and one with 1 hour of data. This means that the default view is going to have a step of 14 seconds in the 1 hour view, and 86 seconds in the 6 hour view.

My first guess with #2059 based on what they're showing is that they've configured a scrape interval of 30 seconds. This is going to lead to the weird +- 50% weird artifacts when you have a miss-matched collection to the scrape. Then when you combine 14 second steps and Prometheus rate extrapolation, you're going to see this +-50% problem.

Basically, they've got a self-induced problem, and it's neither cAdvisors or Prometheus that is the cause.

**Comment by dashpole** [comment]
Another thing that is relevant here is that cAdvisor jitters the interval to spread out load.  Collection occurs every 10-15 seconds.  I'm not entirely sure if that matters for this problem.

**Comment by SuperQ** [comment]
> From my understanding, it is a best practice to perform collection at scrape time, and thus not expose timestamps.

Yes, you're very right about this. The best practice is to collect at scrape time.

> Not attaching timestamps for cached metrics would break rate calculations for all backends

No, this is not likely a problem.

> Collection occurs every 10-15 seconds. I'm not entirely sure if that matters for this problem.

No, this shouldn't be a problem.

**Comment by Wayde2014** [comment]
I also encountered the same confusion, is there a solution to this problem now?

Does cadvisor support turn off timestamps in exported metrics?

@muzammil360

**Comment by muzammil360** [comment]
@Wayde2014 no i didn't find anything else. But i suppose that was like 2 years ago. Cadvisor might have improved considerably during this time. I ended up running a regexp and dropping time field (as far i can remember).

**Comment by muzammil360** [comment]
@Wayde2014 actually, if you look just above your comment, it seems jasonk000 attempted to fix it [here](https://github.com/open-telemetry/opentelemetry-collector/pull/1951)

![image](https://user-images.githubusercontent.com/20393633/153980633-4a4f2e18-8485-46b7-aa90-a99f177024e0.png)

**Comment by Wayde2014** [comment]
> Thanks for the reply.

Do you still remember how the regular expression was written, thank you very much.

below is the command and its execution result:
`$ curl -k -m 5 --header "Authorization: Bearer ${TOKEN}" "https://localhost:10250/metrics/cadvisor"

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0# HELP cadvisor_version_info A metric with a constant '1' value labeled by kernel version, OS version, docker version, cadvisor version & cadvisor revision.
# TYPE cadvisor_version_info gauge
cadvisor_version_info{cadvisorRevision="",cadvisorVersion="",dockerVersion="Unknown",kernelVersion="4.15.0-76-generic",osVersion="Ubuntu 18.04.6 LTS"} 1
# HELP container_cpu_load_average_10s Value of container cpu load average over the last 10 seconds.
# TYPE container_cpu_load_average_10s gauge
container_cpu_load_average_10s{container="",id="/",image="",name="",namespace="",pod=""} 0 1644893651118
container_cpu_load_average_10s{container="",id="/kubepods",image="",name="",namespace="",pod=""} 0 1644893653633
container_cpu_load_average_10s{container="",id="/kubepods/besteffort",image="",name="",namespace="",pod=""} 0 1644893651525
container_cpu_load_average_10s{container="",id="/kubepods/besteffort/pod00b3e4fe-cebd-47a5-8d8c-a2ffbccc146c",image="",name="",namespace="baetyl-edge",pod="idp-web-546d968bb4-4k7fj"} 0 1644893644069
container_cpu_load_average_10s{container="",id="/kubepods/besteffort/pod00b3e4fe-cebd-47a5-8d8c-a2ffbccc146c/5a5a8cce749091bc07596a64b83ade25717d7e67584771010c760d3974c7d104",image="docker.io/rancher/pause@sha256:d22591b61e9c2b52aecbf07106d5db313c4f178e404
d660b32517b18fcbf0144",name="5a5a8cce749091bc07596a64b83ade25717d7e67584771010c760d3974c7d104",namespace="baetyl-edge",pod="idp-web-546d968bb4-4k7fj"} 0 1644893652946
container_cpu_load_average_10s{container="",id="/kubepods/besteffort/pod2bb0af3a-c357-4957-9339-a64112c209f7",image="",name="",namespace="baetyl-edge",pod="redis-7b9c5479b5-ml7qw"} 0 1644893645013
container_cpu_load_average_10s{container="",id="/kubepods/besteffort/pod2bb0af3a-c357-4957-9339-a64112c209f7/0d9a2a747eefa9b74f0b7d3a4179b5a17cb5018bff65d1a7a99293f0d8027826",image="docker.io/rancher/pause@sha256:d22591b61e9c2b52aecbf07106d5db313c4f178e404d660b32517b18fcbf0144",name="0d9a2a747eefa9b74f0b7d3a4179b5a17cb5018bff65d1a7a99293f0d8027826",namespace="baetyl-edge",pod="redis-7b9c5479b5-ml7qw"} 0 1644893648897
container_cpu_load_average_10s{container="",id="/kubepods/besteffort/pod310f1582-3c68-4c8d-9210-f4d02a70193f",image="",name="",namespace="baetyl-edge",pod="rabbitmq-8864f6f8d-fscf7"} 0 1644893650426
container_cpu_load_average_10s{container="",id="/kubepods/besteffort/pod310f1582-3c68-4c8d-9210-f4d02a70193f/37b15f0d73622aceb6ca28e03e30797b194d212a858dc6cdf0dbc8ee0cb51da7",image="docker.io/rancher/pause@sha256:d22591b61e9c2b52aecbf07106d5db313c4f178e404
d660b32517b18fcbf0144",name="37b15f0d73622aceb6ca28e03e30797b194d212a858dc6cdf0dbc8ee0cb51da7",namespace="baetyl-edge",pod="rabbitmq-8864f6f8d-fscf7"} 0 1644893654466
container_cpu_load_average_10s{container="",id="/kubepods/besteffort/pod37594157-6433-4828-b6e6-9e2cc7985516",image="",name="",namespace="kube-system",pod="local-path-provisioner-6d59f47c7-9gl4j"} 0 1644893645231
container_cpu_load_average_10s{container="",id="/kubepods/besteffort/pod37594157-6433-4828-b6e6-9e2cc7985516/af74363cc221281ca74dc33aea1845bd6e4de6cfd4b1381f3b829c7b840e37bc",image="docker.io/rancher/pause:3.1",name="af74363cc221281ca74dc33aea1845bd6e4de6c
fd4b1381f3b829c7b840e37bc",namespace="kube-system",pod="local-path-provisioner-6d59f47c7-9gl4j"} 0 1644893658118
container_cpu_load_average_10s{container="",id="/kubepods/besteffort/pod376f2fed-ba10-419b-9144-11b28a7d9c89",image="",name="",namespace="baetyl-edge-system",pod="baetyl-broker-64b6f574ff-jcbck"} 0 1644893657084
container_cpu_load_average_10s{container="",id="/kubepods/besteffort/pod376f2fed-ba10-419b-9144-11b28a7d9c89/ad4a99c430a3e60aeb4bf93dd8c6ee82bcbad926cd1a160b8e61495caa5e438a",image="docker.io/rancher/pause@sha256:d22591b61e9c2b52aecbf07106d5db313c4f178e404
d660b32517b18fcbf0144",name="ad4a99c430a3e60aeb4bf93dd8c6ee82bcbad926cd1a160b8e61495caa5e438a",namespace="baetyl-edge-system",pod="baetyl-broker-64b6f574ff-jcbck"} 0 1644893643882
container_cpu_load_average_10s{container="",id="/kubepods/besteffort/pod3fa34485-e85b-4de6-b755-fb623a9a4f42",image="",name="",namespace="baetyl-edge",pod="syht-sc-6b7dbbbf64-456dk"} 0 1644893656552
container_cpu_load_average_10s{container="",id="/kubepods/besteffort/pod3fa34485-e85b-4de6-b755-fb623a9a4f42/7319091cdc9ce9baf67648979ebd4d22e0f04fb4129cb77efb2eaa40853fb6a4",image="docker.io/rancher/pause@sha256:d22591b61e9c2b52aecbf07106d5db313c4f178e404
d660b32517b18fcbf0144",name="7319091cdc9ce9baf67648979ebd4d22e0f04fb4129cb77efb2eaa40853fb6a4",namespace="baetyl-edge",pod="syht-sc-6b7dbbbf64-456dk"} 0 1644893652978
container_cpu_load_average_10s{container="",id="/kubepods/besteffort/pod49120d3b-82ad-4844-a923-12b79aeba6d9",image="",name="",namespace="baetyl-edge",pod="idp-service-6595dbf974-zf8pg"} 0 1644893650155
container_cpu_load_average_10s{container="",id="/kubepods/besteffort/pod49120d3b-82ad-4844-a923-12b79aeba6d9/3e92bed31fe20b312d9f2a1c710bbcbbace916ca77b0164ff01c8a88e3681bae",image="docker.io/rancher/pause@sha256:d22591b61e9c2b52aecbf07106d5db313c4f178e404
d660b32517b18fcbf0144",name="3e92bed31fe20b312d9f2a1c710bbcbbace916ca77b0164ff01c8a88e3681bae",namespace="baetyl-edge",pod="idp-service-6595dbf974-zf8pg"} 0 1644893634046
container_cpu_load_average_10s{container="",id="/kubepods/besteffort/pod4e5b65e7-0613-49f5-bc10-f350fa896bec",image="",name="",namespace="baetyl-edge",pod="edge-ota-service-58cb8b9589-s68fd"} 0 1644893649344
container_cpu_load_average_10s{container="",id="/kubepods/besteffort/pod4e5b65e7-0613-49f5-bc10-f350fa896bec/967f6263eb7548aaea840424578675c24cb1327a77143ad3155d0dc929f570d8",image="docker.io/rancher/pause@sha256:d22591b61e9c2b52aecbf07106d5db313c4f178e404
d660b32517b18fcbf0144",name="967f6263eb7548aaea840424578675c24cb1327a77143ad3155d0dc929f570d8",namespace="baetyl-edge",pod="edge-ota-service-58cb8b9589-s68fd"} 0 1644893652032
container_cpu_load_average_10s{container="",id="/kubepods/besteffort/pod5350e932-11d3-4665-885f-2e5e6e456d29",image="",name="",namespace="baetyl-edge",pod="edge-core-iot-575fc6f995-lkg6c"} 0 1644893656366
container_cpu_load_average_10s{container="",id="/kubepods/besteffort/pod5350e932-11d3-4665-885f-2e5e6e456d29/0247a137594b590ebb91555d8192db8b1b5081dda8123a34fef43779ea4e4b7c",image="docker.io/rancher/pause@sha256:d22591b61e9c2b52aecbf07106d5db313c4f178e404
d660b32517b18fcbf0144",name="0247a137594b590ebb91555d8192db8b1b5081dda8123a34fef43779ea4e4b7c",namespace="baetyl-edge",pod="edge-core-iot-575fc6f995-lkg6c"} 0 1644893653306
container_cpu_load_average_10s{container="",id="/kubepods/besteffort/pod561caa04-50b1-4217-a313-24814069066b",image="",name="",namespace="baetyl-edge",pod="minio-754678ff4c-f7f4p"} 0 1644893656948
container_cpu_load_average_10s{container="",id="/kubepods/besteffort/pod561caa04-50b1-4217-a313-24814069066b/fa63882c27b09586c76caa88a6930700a129f8a0ce4a028db821c1d44c7a9268",image="docker.io/rancher/pause@sha256:d22591b61e9c2b52aecbf07106d5db313c4f178e404
d660b32517b18fcbf0144",name="fa63882c27b09586c76caa88a6930700a129f8a0ce4a028db821c1d44c7a9268",namespace="baetyl-edge",pod="minio-754678ff4c-f7f4p"} 0 1644893641717`

**Comment by Wayde2014** [comment]
> @Wayde2014 actually, if you look just above your comment, it seems jasonk000 attempted to fix it [here](https://github.com/open-telemetry/opentelemetry-collector/pull/1951)
> 
> ![image](https://user-images.githubusercontent.com/20393633/153980633-4a4f2e18-8485-46b7-aa90-a99f177024e0.png)

I have seen these, and I have searched online for nearly 2 days, but there is still no good solution. 

Thank you for your enthusiastic help.

**Comment by muzammil360** [comment]
> Thank you for your enthusiastic help.


@Wayde2014 i don't exactly remember. I remember figuring out the pattern of time stamps i wanted to avoid. My regexp was looking for a bunch of digits e.g `1644893653633`. Like you might want to avoid the time from today to 100 years from now. Thsi will limit the no. of most significant digits. 

I must add that i didn't care much about compute complexity for my application. consider for your application.

**Comment by Kiuber** [comment]
@muzammil360 Hi, would you share your regexp rule? thank you.

**Comment by Kiuber** [comment]
`curl ${cadvisor_host}:${cadvisor_host} | sed -r 's/(} .*) ([0-9]*)/\1/'` works for me.

> @muzammil360 Hi, would you share your regexp rule? thank you.

**Comment by Gypsying** [comment]
> `curl ${cadvisor_host}:${cadvisor_host} | sed -r 's/(} .*) ([0-9]*)/\1/'` works for me.
> 
> > @muzammil360 Hi, would you share your regexp rule? thank you.

Thanks, that works for me too.

**Comment by brancz** [comment]
I actually think this can be solved in cadvisor side, since Prometheus supports `honor_timestamp`. Users can that way choose themselves whether they want to honor cadvisor timestamps or rather use the scrape time.

**Comment by dims** [comment]
Closing this issue as it has been open since before January 1, 2024. If this is still relevant, please feel free to reopen or create a new issue.

**Comment by SuperQ** [comment]
FYI for those that find this very old issue.

We added a workaround feature in Prometheus to better handle the internal collection and timestamp data from cAdvisor.

You can now configure the staleness handling at the [scrape_config level](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config)

    track_timestamps_staleness: true

To use this you need to scrape at or faster than the `--housekeeping_interval` to avoid Prometheus missing samples.

For Kubernetes users the default housekeeping interval is hard coded to 10s. This behavior is automatically configured correctly by the [kube-prometheus-stack](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack) chart defaults.

**Comment by dims** [comment]
@SuperQ thank you!
